File: can sum/can_sum_memo.py
# return if any combination of nums can add up to the target num
# nums are greater than 0

# optimized

def can_sum_memo(target, nums, memo={}):
    # checking memo
    # print(target, nums, memo)
    if target in memo:
        return memo[target]

    # base cases
    if target < 0:
        return False
    if target == 0:
        return True

    for num in nums:
        if can_sum_memo(target-num, nums , memo):
            memo[target] = True
            return True
    else:
        memo[target] = False
        return False



# Unlike js, python caches the memo.
# So we should put empty memo to avoid false possitives.
print(can_sum_memo(7,[2, 3], {}))
print(can_sum_memo(7,[5, 3, 4, 7], {}))
print(can_sum_memo(7,[2, 4], {}))
print(can_sum_memo(8,[2, 3, 5], {}))
print(can_sum_memo(300,[7, 14], {}))

File: can sum/can_sum_memo.js
// optimized

const can_sum = (target, nums, memo={}) => {
    console.log(target, nums, memo)
    if (target in memo) return memo[target]
    if (target<0) return false
    if (target===0) return true

    for (let num of nums){
        if (can_sum(target-num, nums, memo) === true){
            memo[target] = true
            return true
        }
    }
    memo[target] = false
    return false
}



// js does not cache the memo.
// So we don't need to put empty memo.
console.log(can_sum(7,[2, 3]))
console.log(can_sum(7,[5, 3, 4, 7]))
console.log(can_sum(7,[2, 4]))
console.log(can_sum(8,[2, 3, 5]))
console.log(can_sum(300,[7, 14]))
